framework/src/service/ServiceReferenceBasePrivate.h

```diff
@@ -124,6 +130,153 @@ namespace cppmicroservices
+     *        The UngetService is called when all instances of the returned shared_ptr object
+     *        go out of scope.
+     */
+    struct UngetHelper
+    {
+        InterfaceMapConstPtr const interfaceMap;
+        ServiceReferenceBase const sref;
+        std::weak_ptr<BundlePrivate> const b;
+
+        UngetHelper(UngetHelper&) = default;
+        UngetHelper(UngetHelper&&) noexcept = default;
+        UngetHelper& operator=(UngetHelper&) = delete;
+        UngetHelper& operator=(UngetHelper&&) noexcept = delete;
+
+        UngetHelper(InterfaceMapConstPtr im, ServiceReferenceBase const& sr, std::shared_ptr<BundlePrivate> const& b)
+            : interfaceMap(std::move(im))
+            , sref(sr)
+            , b(b)
+        {
+        }
+        ~UngetHelper()
+        {
+            try
+            {
+                auto bundle = b.lock();
+                if (sref)
+                {
+                    bool isPrototypeScope
+                        = sref.GetProperty(Constants::SERVICE_SCOPE).ToString() == Constants::SCOPE_PROTOTYPE;
+
+                    if (isPrototypeScope)
+                    {
+                        sref.d.Load()->UngetPrototypeService(bundle, interfaceMap);
+                    }
+                    else
+                    {
+                        sref.d.Load()->UngetService(bundle, true);
+                    }
+                }
+            }
+            catch (...)
+            {
+                // Make sure that we don't crash if the shared_ptr service object outlives
+                // the BundlePrivate or CoreBundleContext objects.
+                if (!b.expired())
+                {
+                    DIAG_LOG(*b.lock()->coreCtx->sink)
+                        << "UngetHelper threw an exception. " << util::GetLastExceptionStr();
+                }
+                // don't throw exceptions from the destructor. For an explanation, see:
+                // https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md
+                // Following this rule means that a FrameworkEvent isn't an option here
+                // since it contains an exception object which clients could throw.
+            }
+        }
+    };
+
+    class CustomServiceDeleter
+    {
+      public:
+        CustomServiceDeleter(ServiceHolder<void>* sh) : sh_(sh), uh_(nullptr) {}
+        CustomServiceDeleter(UngetHelper* uh) : sh_(nullptr), uh_(uh) {}
+
+        void
+        operator()(ServiceHolder<void>* sh)
+        {
+            delete sh;
+        }
+
+        void
+        operator()(UngetHelper* uh)
+        {
+            delete uh;
+        }
+
+        [[nodiscard]] ServiceReferenceBase
+        getServiceRef() const
+        {
+            if (sh_)
+            {
+                return sh_->sref;
+            }
+            return uh_->sref;
+        }
+
+      private:
+        ServiceHolder<void> const* const sh_;
```

Comment:
This is just a matter of coding style and personal preference... but I personally really dislike naming member variables with the trailing underscore. I find that it makes reading code like "ah_->do_something()" with the "_->" awkward to read. If we really need to name a member variable in a special way to mark it as such, I prefer using the "m_ah" form. The code doesn't look strange and the "m_" is very visible... more so than a trailing "_".

For what it's worth :-)